### PR TITLE
Initial work on S7 support for covr

### LINKED ISSUE #578: Add support for S7 methods
Not sure how much work this will be, but hopefully it's just a matter of teaching covr to recognise S7 method registration and add the instrumentation to the right place.

### PR DESCRIPTION
**Pull request by jimhester**
Fixes #578

### CODE REVIEW

**Comment by jimhester** [comment]
This seems to be mostly working, except for the dynamic getter and setter, the lines are showing as uncovered even though they should have been run.

Additionally it doesn't currently handle S7 methods defined on generics outside of the package being covered, or S7 methods for S3 generics.

@hadley I haven't been keeping up with the S7 changes in the past few years, so it is possible I have missed some features. Would love comments or a PR extending the test code to ensure we have handled all known cases.

I have only a limited amount of time to devote to this, but wanted to get it started at least.

**Comment by hadley** [comment]
Awesome, thanks so much for making a start on this!

**Comment by hadley** [comment]
Now that @t-kalinowski has got the checks passing, I'd suggest we merge this and then iterate as we discover bits of S7 that this doesn't cover. This PR is already a big improvement 😄

**Comment by t-kalinowski** [comment]
I have a WIP draft for counting calls to external methods registered with `` S7::`method<-` ``. Still need to write tests for it. I'll open another PR for that.

File: R/S7.R [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      lapply(c("getter", "setter", "validator"), function(prop_fun) {
+        if (!is.null(p[[prop_fun]])) {
+          replacement(
+            sprintf("%s@properties$%s$%s", class_name, p$name, prop_fun),
+            env = p,
+            target_value = p[[prop_fun]])
+        }
+      })
+  })
+  prop_fun_replacements <- unlist(prop_fun_replacements, FALSE, FALSE)
```
Comment (jimhester) [inline R/S7.R:44]:
```suggestion
  prop_fun_replacements <- unlist(prop_fun_replacements, recursive = FALSE, use.names = FALSE)
```

File: R/S7.R [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      target_value <- get(name, envir = env)
+      if (is.environment(target_value)) {
+        # Recurse for nested environments
+        get_replacements(target_value)
+      } else {
+        name <- as.character(attr(target_value, "name", TRUE) %||% name)
+        list(replacement(name, env, target_value))
+      }
+    })
+    unlist(replacements, FALSE, FALSE)
```
Comment (jimhester) [inline R/S7.R:26]:
```suggestion
    unlist(replacements, recursive = FALSE, use.names = FALSE)
```

File: R/S7.R [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  # Each binding in the environment at x@methods is either a function or, for
+  # generics that dispatch on multiple arguments, another environment.
+  get_replacements <- function(env) {
+    replacements <- lapply(names(env), function(name) {
+      target_value <- get(name, envir = env)
+      if (is.environment(target_value)) {
+        # Recurse for nested environments
+        get_replacements(target_value)
+      } else {
+        name <- as.character(attr(target_value, "name", TRUE) %||% name)
```
Comment (jimhester) [inline R/S7.R:22]:
```suggestion
        name <- as.character(attr(target_value, "name", exact = TRUE) %||% name)
```

**Comment by jimhester** [comment]
A few minor style things, but otherwise the improvements look good to me, thanks for pushing this forward!

**Comment by t-kalinowski** [comment]
Thanks! I'll merge this once tests finish running.

**Pull request merged**